### PR TITLE
Add list constructor expression support

### LINKED ISSUE #22357: Add list constructor expression
**Description:**

Refer: https://ballerina.io/spec/lang/2020R1/#list-constructor-expr

### PR DESCRIPTION
**Pull request by lochana-chathura**
Fixes #22357 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 1]
```diff
@@ -6847,4 +6852,39 @@ private STNode parseTrapKeyword() {
...
+     * @return Parsed node
+     */
+    private STNode parseOptionalExpressionsList() {
+        STToken nextToken = peek();
+
+        // Return an empty list if list is empty
+        if (isEndOfListenersList(nextToken.kind)) {
+            return STNodeFactory.createNodeList(new ArrayList<>());
+        }
+        return parseListeners();
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:6888]: I reuse the `parseListeners()` here.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java [hunk 2]
```diff
@@ -261,6 +261,9 @@
     private static final ParserRuleContext[] TYPEDESC_RHS = {ParserRuleContext.NON_RECURSIVE_TYPE,
             ParserRuleContext.ARRAY_TYPE_DESCRIPTOR, ParserRuleContext.OPTIONAL_TYPE_DESCRIPTOR };
 
+    private static final ParserRuleContext[] LIST_CONSTRUCTOR_RHS =
+            { ParserRuleContext.CLOSE_BRACKET, ParserRuleContext.LISTENERS_LIST };
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java:265]: { ParserRuleContext.CLOSE_BRACKET, ParserRuleContext.EXPRESSION }

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java [hunk 3]
```diff
@@ -2297,14 +2309,16 @@ private ParserRuleContext getNextRuleForCloseBrace(int nextLookahead) {
             case FORK_STMT:
                 endContext(); // end fork-statement
                 return ParserRuleContext.STATEMENT;
+            case LIST_CONSTRUCTOR:
+                return ParserRuleContext.CLOSE_BRACKET;
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java:2313]: Can a close-brace be present in a list-constructor context?

Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java:2313]: This is a mistaken addition from a previous commit of mine. I'll fix.

**Pull request merged**